### PR TITLE
Datareaders

### PR DESCRIPTION
**Pull request by flauted**
This PR introduces data readers, originally part of #1194 . The purpose is to further abstract the data type. Specifically, readers make ``inputters.build_dataset`` nearly data type-free.

The key abstraction is that data type instances can be instantiated with data type-specific arguments (e.g. ``normalize_audio`` for the ``AudioDataReader`` or ``channel_size`` for the ``ImageDataReader``) and then ran with the data type-independent ``read`` method.

The logic of the old dataset's ``make_examples`` class method is now the data reader's ``read`` method. The data type dependent arguments are made attrs in the data reader's ``__init__`` method.

Looking ahead: With this, it should be possible to use one Dataset class with args/attrs ``sort_key`` and ``data_type`` rather than subclasses for each data type. I haven't done that yet to keep the diff slightly more manageable.

### CODE REVIEW

**Comment by vince62s** [comment]
looks great, @bpopeters hope you'll have time to review.

Just a question, would that be possible to embed src_reader and tgt_reader in fields ?

File: onmt/inputters/image_dataset.py [hunk 1]
```diff
@@ -16,57 +17,65 @@
...
+
+    def __init__(self, truncate=None, channel_size=3, **kwargs):
+        self.truncate = truncate
+        self.channel_size = channel_size
+
     @staticmethod
     def _check_deps():
         if any([Image is None, transforms is None, cv2 is None]):
-            ImageDataset._raise_missing_dep(
+            ImageDataReader._raise_missing_dep(
```
Comment (bpopeters) [inline onmt/inputters/image_dataset.py]: shouldn't this be a classmethod?

File: onmt/inputters/text_dataset.py [hunk 2]
```diff
@@ -5,31 +5,14 @@
...
-        dynamic_dict (bool)
-    """
-
-    @staticmethod
-    def sort_key(ex):
-        if hasattr(ex, "tgt"):
-            return len(ex.src[0]), len(ex.tgt[0])
-        return len(ex.src[0])
+class TextDataReader(DataReaderBase):
+    def __init__(self, **kwargs):
```
Comment (bpopeters) [inline onmt/inputters/text_dataset.py]: This `__init__` is unnecessary.

File: onmt/inputters/dataset_base.py [hunk 3]
```diff
@@ -2,19 +2,12 @@
...
 from torchtext.vocab import Vocab
 
 
-# several data readers need optional dependencies. There's no
-# appropriate builtin exception
-class MissingDependencyException(Exception):
-    pass
-
-
 class DatasetBase(Dataset):
```
Comment (bpopeters) [inline onmt/inputters/dataset_base.py]: Is `DatasetBase` still an appropriate name for this class? It's just a dataset now, isn't it?

File: onmt/inputters/dataset_base.py [hunk 4]
```diff
@@ -87,13 +80,6 @@ def __init__(self, fields, src_examples_iter, tgt_examples_iter,
...
         super(DatasetBase, self).__init__(examples, fields, filter_pred)
 
-    @staticmethod
-    def _raise_missing_dep(*missing_deps):
-        """Raise missing dep exception with standard error message."""
-        raise MissingDependencyException(
-            "Could not create reader. Be sure to install "
-            "the following dependencies: " + ", ".join(missing_deps))
-
     def __getattr__(self, attr):
```
Comment (bpopeters) [inline onmt/inputters/dataset_base.py]: This method predates this pull request, but what is it for? `self.fields` should be defined from the moment the constructor is done.

**Comment by bpopeters** [comment]
I left some comments inline.

If this makes it easier to process arbitrary pairs of source and target data types, I think it's a very positive development. However, I think there are some additional issues to consider (I intended to make PRs addressing these in the near future).

First, my view is that `make_examples`/`read` should happen inside the constructor of the dataset, not separately beforehand in `inputters.build_dataset`. This is the current behavior and I think it is undesirable -- you wouldn't ever want to call the constructor without first calling `make_examples`/`read` and you wouldn't ever want to call `make_examples`/`read` unless you were about to give it to the dataset constructor. The separation between them is arbitrary.

The second issue flows from this. If the main inputs to the dataset constructor are a source example iterator and a target example iterator, then we force ourselves into a single-source, single-target paradigm. This is unfriendly for things that a lot of practitioners would find useful, such as multisource and multimodal translation. It's often much more convenient to store this sort of data in a single table than in separate files for the source, target, and whatever other data you have. This is a concrete issue I'm having with one of my current research projects. There's nothing intrinsically difficult about processing tabular data with torchtext (see [here](https://github.com/pytorch/text/blob/master/torchtext/data/dataset.py#L217)), but I want to make sure that none of our current work makes it more complicated than it needs to be. I think the extension is easier if we adopt the structure I suggest in my first point.

@vince62s I don't think that would be a desirable change. How the data is read/iterated over is independent of how it should be tokenized and numericalized, so they should not be combined.

File: onmt/inputters/dataset_base.py (hunk 4 above)
Comment (flauted) [reply onmt/inputters/dataset_base.py]: (Specifically it's from #1188) Those two lines are necessary. It's for the unpickling side. Pickle does a ``getattr`` call to try to find ``__setstate__``. Since the state is not yet set, ``self.fields`` actually isn't defined. I got it from [here](https://stackoverflow.com/a/49380669/8217456).

File: onmt/inputters/dataset_base.py (hunk 3 above)
Comment (flauted) [reply onmt/inputters/dataset_base.py]: I agree but then IMO the file needs renamed. I didn't want to blow up the diff by doing that now since there are actual changes to review here.

File: onmt/inputters/text_dataset.py (hunk 2 above)
Comment (flauted) [reply onmt/inputters/text_dataset.py]: I've designed it so that any reader can take the union of all the reader args, if that makes sense. So the ``**kwargs`` is doing something. That way in translator.py ``Translator.__init__`` and in preprocess.py ``main`` you don't have to do logic on the datatype to decide what arguments to pass in to the reader. You just pass them all in and whatever subclass takes what it wants and ignores the rest.

I would say that should be in the ``DataReaderBase`` "contract." However, putting the empty ``__init__`` in the base class suggests that subclasses need to call super init, and that's not the case.

Do you have something else in mind?

File: onmt/inputters/image_dataset.py (hunk 1 above)
Comment (flauted) [reply onmt/inputters/image_dataset.py]: Yeah probably.

**Comment by flauted** [comment]
Yeah my end use case also runs up on the second issue. Captioning datasets often have multiple target captions per "source" video. In some datasets (MSVD, aka youtube2text) that number of target captions is actually variable. Usually the strategy is to train each video-caption pair, repeating the different videos. When I did a prototype of adding video captioning to onmt, my solution was to "flatten" the captions and make symlinks to repeat the videos ahead of time. That's not elegant.

FWIW, I haven't played with the TabularDataset at all, but usually "tabular" solutions (I'm thinking Pandas) don't play well with the variable number of captions per video.

On your first issue: sharding calls ``build_dataset`` multiple times with an expectation that the data is read the same way for each shard (and it's the same expectation for train/val currently). That's why I think having a "reader" abstraction that's independent of the dataset makes a lot of sense. I would suggest having the reader be passed as an argument to the dataset constructor and doing the ``read``ing within. I can add that to this PR if you think it's a good idea.

**Comment by bpopeters** [comment]
I agree that the reader abstraction makes sense. Having the readers be arguments of the dataset constructor is a good idea.

My comments about tabular data were not specific to any particular tool used to read them in and I doubt any such tool would be necessary. I hadn't considered the possibility of a variable number of inputs, but that sounds like it will be a problem no matter what format the data is in.

File: onmt/inputters/text_dataset.py (hunk 2 above)
Comment (bpopeters) [reply onmt/inputters/text_dataset.py]: I understand the reasoning but it sounds like it would make more sense on the `DataReaderBase` -- otherwise, additional data readers will need to define this dummy constructor as well. Could you add a comment so no one makes the same mistake as I did?

File: onmt/inputters/dataset_base.py (hunk 4 above)
Comment (bpopeters) [reply onmt/inputters/dataset_base.py]: Ah, understood.

**Comment by vince62s** [comment]
it's actually what I meant to avoid these src_reader arguments everywhere.
But, couldn't we deduct the reader from the data_type in the dataset_constructor ?

**Comment by flauted** [comment]
@vince62s Maybe. The problem is, DataReader's are meant to be instantiated once. Not for each shard for each split (like the Dataset is). So you would have to instantiate the readers you need outside of the each shard/each split loop. Then you have to cache the instances somewhere so that ``build_dataset`` can bring them in. That introduces a pretty weird runtime state. Passing the args seems conceptually cleaner to me.

File: onmt/inputters/text_dataset.py (hunk 2 above)
Comment (flauted) [reply onmt/inputters/text_dataset.py]: Done.

**Comment by vince62s** [comment]
ok I misunderstood both of you. I thought dataset attributes not dataset arguments.
I am wondering why we don't pass only reader_args and then the Datareaders are Datasets' attributes.
Wouldn't that make things simpler ?

**Comment by flauted** [comment]
Do you mean: at runtime assign the data reader instances as a class attribute to DatasetBase?

You could. That's one way of doing what I said above: caching the instances somewhere (in the class state), creating a weird state ("weird" in the sense of a class having a state).

It becomes awkward if, say, a library user makes a script that makes (src=text, tgt=text) datasets and then makes (src=text, tgt=img) datasets. You have to set the DatasetBase state once, make the first type of datasets, and then set the DatasetBase state again to make the second type of datasets. That example is contrived, but it illustrates a downside of the class-state solution.

If you and @bpopeters think that doing that sort of stateful solution is really better than passing some additional arguments, then I can try to implement it. Just let me know.

**Comment by vince62s** [comment]
> 
> It becomes awkward if, say, a library user makes a script that makes (src=text, tgt=text) datasets and then makes (src=text, tgt=img) datasets. You have to set the DatasetBase state once, make the first type of datasets, and then set the DatasetBase state again to make the second type of datasets. That example is contrived, but it illustrates a downside of the class-state solution.
> 

I don't understand what you describe above and how the current PR would make things "smoother".

**Comment by bpopeters** [comment]
> Do you mean: at runtime assign the data reader instances as a class attribute to DatasetBase?

I agree with you that this is not a good idea. I hope none of my previous comments seemed to endorse it. I was suggesting the readers should be arguments to the dataset constructor. Whether they also need to be attributes of a dataset _instance_ after that -- I'm not sure. But definitely not class attributes.

**Comment by bpopeters** [comment]
It one _were_ to use class attributes for the readers, then the natural thing to do would be to define a separate derived dataset class for each combination of readers. But I think that's inferior to not messing with class attributes in the first place.

This most recent commit gets around that, but yes, class attributes are a very strange way to solve the problem. We are in agreement about this.

**Comment by flauted** [comment]
>I am only saying it does not necessarily makes things easier from the top level (preprocess / translator) point of view.

This is fair criticism. Would it help to define a ``build_readers`` function that makes the readers from the opts: ```readers = build_readers(opt)```?

>I guess it was more what Ben is saying "attributes of the instance".

I'm not sure what that buys us.

>I also don's how it is goinf to make things easier to accept multi source.

At the preprocess/translator level, this PR doesn't. The only changes that'll make multi source easier are at the ``DatasetBase.__init__`` level. It now takes an arbitrary number of readers and arguments to ``read`` instead of ``src_examples_iter, tgt_examples_iter``.

**Comment by vince62s** [comment]
Don't do it, let's just discuss:
What I was thinking, is to move the code src/tgt_reader = inputters.str2reader from the top level scipts into the inputters.build_dataset() function and to pass only one argument readers_arg to the inputters.build_dataset function.
I understand this would execute the code for each shard but is this an issue ?

**Comment by vince62s** [comment]
I'm fine with 1) or 2). @bpopeters what do you think ? (btw would that be another opportunity to use the syntax readers.from_opt() as it was for optim ? )

**Comment by flauted** [comment]
Oh yeah. Having a ``from_opt`` class method on each reader is a good idea. I actually did that on #1194 . Then there's no need for the readers to take the union of all the reader arguments, and there's no need for the dummy ``__init__``. I'll add that now.

**Comment by vince62s** [comment]
@bpopeters  do we have your blessing on this PR ?

**Comment by bpopeters** [comment]
It looks good to me.

**Comment by vince62s** [comment]
In our tests of FP16 on master we discovered several small bugs that were fixed.
I would like to finish these tests and release a stable version before introducing this PR.
Since we lack some unit tests, I don't want to take the risk to break too much at the same time.
btw thanks @guillaumekln for fixing the previous bugs.

**Comment by flauted** [comment]
Now that 0.7.2 is tagged, is this waiting on anything else @vince62s ?

**Comment by vince62s** [comment]
ok, merging.
